### PR TITLE
Enable //brave/browser/ui deps checks.

### PR DESCRIPTION
**Pull request by goodov**
What's changed here:

1. Make `//chrome/browser/ui` publicly depend on `//brave/browser/ui`, allow `//brave/browser/ui` to include files from `//chrome/browser/ui`.
2. Forbid `//brave/browser/ui` dependency, use `//chrome/browser/ui` instead.
3. Make `//chrome/browser` privately depend on `//brave/browser/ui`, allow `//brave/browser/ui` to include files from `//chrome/browser`. This is similar to how `//chrome/browser/ui` target [works upstream](https://github.com/chromium/chromium/blob/589b22b1f47883ded56c9c2776c23a5445374197/chrome/browser/BUILD.gn#L1910).

Windows x64 test release build is [successful](https://ci.brave.com/job/test-brave-browser-build-windows-x64/807/).
Android test release builds are [successful](https://ci.brave.com/job/test-brave-browser-build-android/195/).

This PR doesn't try to migrate `//brave/browser/ui` to `sources.gni` approach to inject everything directly into `//chrome/browser/ui`, but this can be done as a next step.

Resolves https://github.com/brave/brave-browser/issues/28912

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/BUILD.gn [hunk 1]
```diff
@@ -164,14 +164,16 @@ static_library("browser") {
...
+  public_deps = [
     ":constants",
     ":generated_json_rpc_responses",
     ":generated_swap_responses",
     ":hd_keyring",
     ":pref_names",
     ":transaction",
     ":utils",
+  ]
+  deps = [
```
Comment (darkdh) [inline components/brave_wallet/browser/BUILD.gn:176]: Why is this needed? If a target wants to depend on ex. `constants`, it will put `//brave/components/brave_wallet/browser:constants` into its dep instead of `//brave/components/brave_wallet/browser` to get all public deps implicitly.

**Review by simonhong — APPROVED** [review]
++

**Comment by goodov** [comment]
@brave/rewards-client PTAL
@brave/patch-reviewers PTAL

**Comment by goodov** [comment]
@sangwoo108 I'd like to hear your inputs/see an approve, your UI experience is very valuable here!

Previously we had `//brave/browser/ui` in a state "it compiles somehow, but can break at any point". Now with added dependency checks and a fixed build tree we should have it compile in most cases as expected.

File: browser/ui/DEPS [hunk 2]
```diff
@@ -1,4 +1,22 @@
...
+
+  # Restrict views similar to chrome/browser/ui/DEPS.
+  "-brave/browser/ui/views",
+  "-chrome/browser/ui/views",
 ]
+
+specific_include_rules = {
+  "browser_commands\.cc": [
+    "+brave/browser/ui/views",
+  ],
```
Comment (sangwoo108) [inline browser/ui/DEPS:17]: Is this for @fallaciousreasoning 's work? Would it be needed even when @fallaciousreasoning refactors the code layout as you suggested(https://github.com/brave/brave-core/pull/17296#discussion_r1127313638) ?

**Review by sangwoo108 — APPROVED** [review]
Really impressive work! LGTM with a small question 👍

File: browser/ui/DEPS (hunk 2 above)
Comment (goodov) [reply browser/ui/DEPS:17]: > Is this for @fallaciousreasoning 's work? Would it be needed even when @fallaciousreasoning refactors the code layout as you suggested([#17296 (comment)](https://github.com/brave/brave-core/pull/17296#discussion_r1127313638)) ?

yep, it's a temporary thing. We should remove this in his refactoring PR.

**Review by emerick — APPROVED** [review]
Rewards changes LGTM

File: components/brave_wallet/browser/BUILD.gn (hunk 1 above)
Comment (darkdh) [reply components/brave_wallet/browser/BUILD.gn]: >That's correct, but do we really want to micro-manage all these sub-dependencies when we add //brave/components/brave_wallet/browser dependency?

I don't think `public_deps` is designed for eliminating these sub deps, it is used for ex. target A depends on `ui` which depends on `//brave/components/brave_wallet/browser` and A couldn't depends on `//brave/components/brave_wallet/browser:constants` due to visibility but want to include headers from it.
It's usually used for `components`, `static_library`, `shard_library` or some targets that have visibility restriction.

**Review by darkdh — APPROVED** [review]

**Review by fallaciousreasoning — APPROVED** [review]
LGTM, thanks @goodov  :smile:

File: components/brave_wallet/browser/BUILD.gn (hunk 1 above)
Comment (goodov) [reply components/brave_wallet/browser/BUILD.gn]: > It's usually used for `components`, `static_library`, `shard_library` or some targets that have visibility restriction.

I'd like to point out another `public_deps` use case for anyone reading this thread: when a target uses generated headers in its own public headers, such as `mojom-forward.h`, we should use `public_deps += [ ":mojom" ]`, otherwise the build might fail occasionally if an upstream target doesn't depend on `:mojom` explicitly.

So `public_deps` is not only about visibility, but also about propagating the correct build graph dependency to upstream targets so ninja knows not to build some files until some dependencies are ready.

**Review by iefremov — APPROVED** [review]
